lucene/core/src/java/org/apache/lucene/codecs/blocktree/CompressionAlgorithm.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.codecs.blocktree;
+
+import java.io.IOException;
+
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.util.compress.LowercaseAsciiCompression;
+
+/**
+ * Compression algorithm used for suffixes of a block of terms.
+ */
+enum CompressionAlgorithm {
+
+  NO_COMPRESSION(0x00) {
+
+    @Override
+    void read(DataInput in, byte[] out, int len) throws IOException {
+      in.readBytes(out, 0, len);
+    }
+
+    @Override
+    public String toString() {
+      return "no_compression";
+    }
+  },
+
+  LOWERCASE_ASCII(0x01) {
+
+    @Override
+    void read(DataInput in, byte[] out, int len) throws IOException {
+      LowercaseAsciiCompression.decompress(in, out, len);
+    }
+
+    @Override
+    public String toString() {
+      return "lowercase_ascii";
+    }
+  },
+
+  LZ4(0x02) {
+
+    @Override
+    void read(DataInput in, byte[] out, int len) throws IOException {
+      org.apache.lucene.util.compress.LZ4.decompress(in, len, out, 0);
+    }
+
+    @Override
+    public String toString() {
+      return "lz4";
+    }
+  };
+
+  private static final CompressionAlgorithm[] BY_CODE = new CompressionAlgorithm[3];
```

Comment:
Why bother with this -- explicit "code" in constructors (when the enum's intrinsic ordinal will do) and explicit BY_CODE construction when you could do `private static final CompressionAlgorithm[] BY_CODE = values();`

Reply:
I'm fine either way but I know some people strongly prefer decoupling ids - which are used for serialization and shouldn't change - from enum ordinals that might change if new constants are inserted or if values are reordered.

Reply:
I understand that but it sandbags the current implementation with verbosity it doesn't need.  We can adjust in the future if we actually need to.

Reply:
Exactly. I prefer the explicit code. Gives you control and context. Ordinals are really internal details of a particular implementation of enums.

Reply:
I don't view the public methods of an Enum as "internals".

Reply:
BTW I can understand not wanting clients of an enum to call ordinal() on it; they should call a `getCode()` method, and the implementation of that could be ordinal().

Reply:
You are entitled to your opinion, David. But the comment Adrien made still holds: when you reorder or add/ remove enum constants, their ordinal will change and this side effect is worth guarding against. The "code" may be verbose but it is also explicit and makes making a mistake more difficult. I think it's valuable here.

Anyway. My original comment about hardcoded values was only to point out that code used the same constants in different places. I suggested an enum (like Adrien implemented) but it could as well be a set of constant integers declared in one place. I don't care about this (but I do care about using or overriding ordinal()...).

Reply:
It doesn't address the concern that something that looks as innocuous as reordering constants or adding one at any position but in the end breaks serialization. I have a slight preference for being explicit like Dawid, but I've been in the middle of this controversy many times already and it's a matter of robustness vs. conciseness in the end. If you feel strongly about it @dsmiley I'll switch back to using `#ordinal()`.

Reply:
Oops I had not seen @dweiss 's comment when writing mine.

Reply:
+1 for the explicit codes too.  Relying on enum ordinals is dangerously fragile ...